internal/controllers/core/portforward/reconciler_test.go

```diff
@@ -215,57 +218,81 @@ func TestMultipleForwardsMultiplePods(t *testing.T) {
 	require.True(t, seenFoo, "did not see port forward foo")
 	require.True(t, seenBar, "did not see port forward bar")
 
-	state = f.st.LockMutableStateForTesting()
-	delete(state.PortForwards, pfFooName)
-	f.st.UnlockMutableState()
+	f.Delete(pfFoo)
 
-	f.onChange()
+	f.wait()
 	require.Equal(t, 1, len(f.r.activeForwards))
 	f.assertContextCancelled(t, ctxFoo)
 	f.assertContextNotCancelled(t, ctxBar)
 }
 
 type pfrFixture struct {
-	*tempdir.TempDirFixture
-	ctx    context.Context
-	cancel func()
-	kCli   *k8s.FakeK8sClient
-	st     *store.TestingStore
-	r      *Reconciler
-	out    *bufsync.ThreadSafeBuffer
+	*fake.ControllerFixture
+	t       *testing.T
+	ctx     context.Context
+	cancel  func()
+	kCli    *k8s.FakeK8sClient
+	ctrlCli ctrlclient.Client
+	st      *store.TestingStore
+	r       *Reconciler
+	out     *bufsync.ThreadSafeBuffer
 }
 
 func newPFRFixture(t *testing.T) *pfrFixture {
-	f := tempdir.NewTempDirFixture(t)
 	st := store.NewTestingStore()
 	kCli := k8s.NewFakeK8sClient(t)
-	r := NewReconciler(kCli)
-
+	ctrlCli := fake.NewTiltClient()
+	r := NewReconciler(st, kCli, ctrlCli)
+	cf := fake.NewControllerFixture(t, r)
 	out := bufsync.NewThreadSafeBuffer()
 	ctx, cancel := context.WithCancel(context.Background())
 	ctx = logger.WithLogger(ctx, logger.NewTestLogger(out))
 	return &pfrFixture{
-		TempDirFixture: f,
-		ctx:            ctx,
-		cancel:         cancel,
-		st:             st,
-		kCli:           kCli,
-		r:              r,
-		out:            out,
+		ControllerFixture: cf,
+		t:                 t,
+		ctx:               ctx,
+		cancel:            cancel,
+		st:                st,
+		kCli:              kCli,
+		ctrlCli:           ctrlCli,
+		r:                 r,
+		out:               out,
 	}
 }
 
-func (f *pfrFixture) onChange() {
-	f.r.OnChange(f.ctx, f.st, store.LegacyChangeSummary())
+func (f *pfrFixture) wait() {
```

Comment:
TBH I think the tests are likely to suffer from timing failures from this. I think it might be better to replace this method with something like `requirePfStarted(name string)` which uses `requireState()` to check that `!pf.Status.StartedAt.IsZero()`; i.e. use it as control flow/synchronization and then do normal assertions after.

Reply:
hm yeah given that i haven't yet implemented status yet, i'm just doing "you can get the port forward back from the API server" which is maybe not THE most robust but better than the existing, wdyt?